### PR TITLE
Added a global filter menu for runs

### PR DESCRIPTION
**Pull request by webhead3893**

### CODE REVIEW

File: components/FilterComponent.vue [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+  get filterLabels() {
+    let options = [{index: -1, label: "Global"}];
+    this.parsedSplits.Run.Segments.Segment.forEach((split, index) => options.push({index: index, label: split.Name}));
+
+    return options;
+  }
+
+  get allowActivate() {
+    if (this.filterData.details?.index == -2 || this.filterData.details?.index == undefined)
```
Comment (Webcretaire) [inline components/FilterComponent.vue:76]: I see this `-2` a lot, it should probably be a named constant somewhere (in filter.ts probably), same for `-1` to indicate "Global"

File: components/FilterComponent.vue [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  parsedSplits!: SplitFile;
+
+  useRealTime: boolean = store.state.useRealTime;
+
+  globalFilters: Filter[] = store.state.filters;
+
+  filterData: Filter = {details: {label: "", index: -2}, timeMin: 0, timeMax: 0, active: false, attempts: []};
+
+  get filterLabels() {
+    let options = [{index: -1, label: "Global"}];
```
Comment (Webcretaire) [inline components/FilterComponent.vue:69]: Can be `const` : pushing in an array is still allowed (only `=` is forbidden on a `const` thing)

File: components/FilterComponent.vue [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      } else if (index > -1) {
+        return this.filterListSplits(index);
+      }
+    }
+
+    return [];
+  }
+
+  filterListGlobal() {
+    let filteredAttempts: number[] = [];
```
Comment (Webcretaire) [inline components/FilterComponent.vue:99]: Can be `const` too

File: components/FilterComponent.vue [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+          filteredAttempts.push(attempt['@_id']);
+        }
+      }
+    });
+
+    return filteredAttempts;
+  }
+
+  filterListSplits(index: number) {
+    let filteredAttempts: number[] = [];
```
Comment (Webcretaire) [inline components/FilterComponent.vue:120]: Can be `const` too

File: components/TabsContainer.vue [hunk 5]
```diff
@@ -2,7 +2,7 @@
   <b-card class="main-card" no-body>
     <b-tabs class="main-tabs" card pills align="center" lazy>
       <b-tab title="Summary" active>
-        <div v-if="attemptsExist">
+        <div v-if="parsedSplits.Run.AttemptHistory">
```
Comment (Webcretaire) [inline components/TabsContainer.vue:5]: Is this change intended ? You updated `attemptsExist` but you're not using it anymore

File: components/FilterComponent.vue [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        if (secondsAttemptTime > timeMin && secondsAttemptTime < timeMax && attempt['@_id'] > 0) {
+          filteredAttempts.push(attempt['@_id']);
+        }
+      }
+    });
+
+    return filteredAttempts;
+  }
+
+  filterListSplits(index: number) {
```
Comment (Webcretaire) [inline components/FilterComponent.vue:119]: This looks a lot like `filterListGlobal`, maybe there is a way to pass an `OptionalRealAndGameTime` array to a unique function that works for both

File: components/AttemptSelector.vue [hunk 7]
```diff
@@ -46,10 +46,17 @@ export default class AttemptSelector extends Vue {
   }
 
   get runAttempts(): Attempt[] {
```
Comment (Webcretaire) [inline components/AttemptSelector.vue:48]: Maybe a few comments could be useful to clarify where each "filter" comes from, since there are quite a few here

**Review by Webcretaire — COMMENTED** [review]
Haven't tested yet, just read the code quickly, maybe more comments to come

File: components/TabsContainer.vue (hunk 5 above)
Comment (webhead3893) [reply components/TabsContainer.vue:5]: when filters are active, attemptsExist checks that AttemptHistory exists and that there are attempts in the filtered list, but since none of the filters affect the summary tab, the v-if is just AttemptHistory so that it only gets disabled if there no attempts in the file and not when all are filtered out
there's probably a prettier way to do it but this seemed the most simple

also attemptsExist is used for all of the other tabs, but that's probably not visible in the commit

Comment (Webcretaire) [reply components/TabsContainer.vue:5]: Fair, fair, thanks for explanation, you're right

File: util/filter.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
+export const Global = -1;
```
Comment (Webcretaire) [inline util/filter.ts:1]: Small detail, but in most languages the convention for global constants is screaming snake case:

```ts
export const FILTER_GLOBAL = -1;

export const FILTER_DEFAULT_INDEX = -2;
```

(change usages too of course)

File: components/floating_buttons/FilterRuns.vue [hunk 9]
```diff
@@ -0,0 +1,59 @@
+<template>
+  <b-button v-b-tooltip.hover.left="tooltip" pill size="lg" :variant="variant" @click="filterRuns"
```
Comment (Webcretaire) [inline components/floating_buttons/FilterRuns.vue:2]: Time selectors are a bit too small right now
```suggestion
  <b-button v-b-tooltip.hover.left="tooltip" pill size="xl" :variant="variant" @click="filterRuns"
```

File: components/modals/FilterModal.vue [hunk 10]
```diff
@@ -0,0 +1,34 @@
+<template>
+  <b-modal :ref="modalRef" title="Filter runs" class="text-center" @hidden="destroyModal" hide-footer centered
+           size="lg">
```
Comment (Webcretaire) [inline components/modals/FilterModal.vue:3]: Time selectors are a bit too small right now
```suggestion
           size="xl">
```

File: components/FilterComponent.vue [hunk 11]
```diff
@@ -0,0 +1,147 @@
+<template>
+  <div class="text-center">
+    <div v-if="!filterData.active">
```
Comment (Webcretaire) [inline components/FilterComponent.vue:3]: Since the whole component is a big `if` else it should probably be 2 different components

File: components/AttemptDetailTab.vue [hunk 12]
```diff
@@ -38,13 +40,23 @@ export default class AttemptDetailTab extends Vue {
...
+      attempts = [...this.parsedSplits.Run.AttemptHistory.Attempt].filter(a => this.filteredAttempts.includes(a['@_id']));
+    else
+      attempts = [...this.parsedSplits.Run.AttemptHistory.Attempt];
+
     return attempts.reverse().slice(0, this.attemptsToRender);
   }
 
   @Watch('attemptsLatestToOldest', {immediate: true})
   updateLastsplitByAttempt() {
+    console.log("lol");
```
Comment (Webcretaire) [inline components/AttemptDetailTab.vue:59]: lol

Comment (webhead3893) [reply components/AttemptDetailTab.vue:59]: 😳

File: components/ActiveFilter.vue [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+export default class ActiveFilter extends Vue {
+  @Prop()
+  filterIndex!: number;
+
+  filterData: Filter = store.state.filters[this.filterIndex - 1];
+
+  globalFilters: Filter[] = store.state.filters;
+
+  get displayText() {
+    if (this.filterData.timeMin != undefined && this.filterData.timeMax != undefined && this.filterData.details != undefined)
```
Comment (Webcretaire) [inline components/ActiveFilter.vue:30]: There is no `else` so this getter can return `undefined` potentially. I don't know if it's a big deal but it's a bit weird

File: components/FilterMenu.vue [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+          filteredAttempts.push(run['@_id']);
+        }
+      }
+    });
+
+    return filteredAttempts;
+  }
+
+  activateFilter() {
+    if (this.filterData.timeMin && this.filterData.timeMax){
```
Comment (Webcretaire) [inline components/FilterMenu.vue:105]: That's some complex code, I think 
```ts
if (this.filterData?.timeMin > this.filterData?.timeMax)
    [this.filterData.timeMin, this.filterData.timeMax] = [this.filterData.timeMax, this.filterData.timeMin];
```
would do the same (although can timeMin and timeMax really be undefined ? I think even your default values for filters set them to 0)

**Comment by Webcretaire** [comment]
Very nice, thanks for PR
